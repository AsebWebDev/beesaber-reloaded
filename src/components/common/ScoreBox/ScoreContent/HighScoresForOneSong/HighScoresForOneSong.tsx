import { formatDistance, subDays } from 'date-fns';
import { MDBBadge, MDBIcon, MDBTooltip } from 'mdb-react-ui-kit';
import styled from 'styled-components';

import DiffTag from '@/components/common/DiffTag/DiffTag';
import { parseSongPicUrl } from '@/helper/urlParser';
import tokens from '@/tokens';

import type { Score } from '@/sharedTypes';
// import HighscoreTable from "./HighscoreTable";

const Badge = styled(MDBBadge)`
  color: ${tokens.color.page.bgColor}!important;
  min-width: 30px;
`;

const Img = styled.img`
  max-width: 50px;
  max-height: 30px;
  border-radius: 20px;
`;

const SongData = styled.div`
  display: flex;
  align-items: center;
  gap: 1rem;
`;

type Props = {
  highscore: Score;
};

function HighScoresForOneSong({ highscore }: Props): JSX.Element | null {
  const {
    difficulty,
    levelAuthorName,
    playedBy,
    playedByHive,
    rank,
    score,
    scoreId,
    songAuthorName,
    songHash,
    songName,
    timeSet,
  } = highscore;
  const logid = (id: string) => console.log('id: !', id);

  let scoreColorClass = '';

  if (playedByHive === true && playedBy !== undefined) {
    scoreColorClass =
      playedBy.filter((bee) => bee.myScore < bee.beeScore).length > 0
        ? 'played-by-hive-lost'
        : 'played-by-hive-won';
  }

  return (
    <tr key={scoreId} className={scoreColorClass}>
      <td>
        <Badge>{rank}</Badge>
      </td>
      <td>
        <SongData>
          <DiffTag difficulty={difficulty} />
          <Img src={parseSongPicUrl(songHash)} alt={`Cover of ${songName}`} />
          {playedByHive === true && (
            <MDBTooltip domElement clickable tag="span" placement="top">
              <span>
                <MDBIcon far icon="handshake" />
              </span>
              <div className="also-played-by">
                <table>
                  <thead>
                    <tr>
                      <th>Bee</th>
                      <th>Score</th>
                      <th>My Score</th>
                    </tr>
                  </thead>
                  <tbody id="bees">
                    {/* {playedBy.map((bee, i) => (
                      <HighscoreTable bee={bee} key={i} />
                    ))} */}
                  </tbody>
                </table>
              </div>
            </MDBTooltip>
          )}
          <MDBBadge onClick={() => logid(songHash)} color="dark">
            {songAuthorName} - {songName}
            <span className="greyed-out"> by {levelAuthorName}</span>
          </MDBBadge>
        </SongData>
      </td>
      <td>
        <MDBBadge id="badge-light">{score}</MDBBadge>
      </td>
      <td className="time">
        <b className="card-link">
          {formatDistance(subDays(new Date(), 3), new Date(timeSet), {
            addSuffix: true,
          })}
        </b>
      </td>
    </tr>
  );
}

export default HighScoresForOneSong;
