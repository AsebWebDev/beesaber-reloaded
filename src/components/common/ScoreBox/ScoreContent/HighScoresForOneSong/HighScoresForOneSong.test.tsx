import { render, screen } from '@testing-library/react';

import exampleScores from '@/testing/testData/exampleScores';

import HighScoresForOneSong from './HighScoresForOneSong';

import type { Score } from '@/sharedTypes';

describe('components/common/ScoreBox/ScoreContent/HighScoresForOneSong', () => {
  const highscore = exampleScores[0];
  const highscorePlayedByHive = exampleScores[2];

  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2021-12-01').getTime());
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it.each`
    score                    | description
    ${highscore}             | ${'not played by hive'}
    ${highscorePlayedByHive} | ${'played by hive'}
  `(
    'should match snapshot is $description',
    async ({ score }: { score: Score }) => {
      render(
        <table>
          <tbody>
            <HighScoresForOneSong highscore={score} />
          </tbody>
        </table>
      );

      const component = await screen.findByTestId('highscore-for-one-song');

      expect(component).toMatchSnapshot();
    }
  );
});
