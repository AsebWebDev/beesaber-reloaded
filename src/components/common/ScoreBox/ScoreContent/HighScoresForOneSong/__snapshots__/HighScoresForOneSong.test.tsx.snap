// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/common/ScoreBox/ScoreContent/HighScoresForOneSong should match snapshot is not played by hive 1`] = `
.c2 {
  margin: 1px 2px 1px 2px;
  max-height: 20px;
  max-width: -webkit-fit-content;
  max-width: -moz-fit-content;
  max-width: fit-content;
  min-width: 50px;
}

.c0 {
  color: #00111f!important;
  min-width: 30px;
}

.c3 {
  max-width: 50px;
  max-height: 30px;
  border-radius: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

<tr
  class=""
  data-testid="highscore-for-one-song"
>
  <td>
    <span
      class="badge bg-primary c0"
    >
      2
    </span>
  </td>
  <td>
    <div
      class="c1"
    >
      <span
        class="c2 badge badge-danger"
      >
        Expert+
      </span>
      <img
        alt="Cover of It has Begun"
        class="c3"
        src="https://new.scoresaber.com/api/static/covers/32C3A0E1FEE9FD589B795941B6D79AFE8061D30A.png"
      />
      <span
        class="badge bg-dark"
      >
        Starset
         - 
        It has Begun
        <span
          class="greyed-out"
        >
           by 
          Joshabi
        </span>
      </span>
    </div>
  </td>
  <td>
    <span
      class="badge bg-primary"
      id="badge-light"
    >
      1000
    </span>
  </td>
  <td
    class="time"
  >
    <b
      class="card-link"
    >
      in 16 days
    </b>
  </td>
</tr>
`;

exports[`components/common/ScoreBox/ScoreContent/HighScoresForOneSong should match snapshot is played by hive 1`] = `
.c2 {
  margin: 1px 2px 1px 2px;
  max-height: 20px;
  max-width: -webkit-fit-content;
  max-width: -moz-fit-content;
  max-width: fit-content;
  min-width: 50px;
}

.c0 {
  color: #00111f!important;
  min-width: 30px;
}

.c3 {
  max-width: 50px;
  max-height: 30px;
  border-radius: 20px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

<tr
  class="played-by-hive-won"
  data-testid="highscore-for-one-song"
>
  <td>
    <span
      class="badge bg-primary c0"
    >
      2
    </span>
  </td>
  <td>
    <div
      class="c1"
    >
      <span
        class="c2 badge badge-danger"
      >
        Expert+
      </span>
      <img
        alt="Cover of Let us make a deal"
        class="c3"
        src="https://new.scoresaber.com/api/static/covers/32C3A0E1FEE9FD589B795941B6D79AFE8061D30B.png"
      />
      <span>
        <i
          class="far fa-handshake"
        />
      </span>
      <span
        class="badge bg-dark"
      >
        Stray from the Path
         - 
        Let us make a deal
        <span
          class="greyed-out"
        >
           by 
          Joshabi
        </span>
      </span>
    </div>
  </td>
  <td>
    <span
      class="badge bg-primary"
      id="badge-light"
    >
      2000
    </span>
  </td>
  <td
    class="time"
  >
    <b
      class="card-link"
    >
      in 10 months
    </b>
  </td>
</tr>
`;
